pkg/agent/nodeportlocal/portcache/port_table_windows.go

```diff
@@ -0,0 +1,252 @@
+		}
+		protocolSocketData.State = stateClosed
+	}
+	return nil
+}
+
+func (pt *PortTable) addRuleforFreePort(podIP string, podPort int, protocol string) (int, []ProtocolSocketData, error) {
+	klog.V(2).InfoS("Looking for free Node port on Windows", "podIP", podIP, "podPort", podPort, "protocol", protocol)
+	numPorts := pt.EndPort - pt.StartPort + 1
+	for i := 0; i < numPorts; i++ {
+		port := pt.PortSearchStart + i
+		if port > pt.EndPort {
+			// handle wrap around
+			port = port - numPorts
+		}
+		if _, ok := pt.NodePortTable[port]; ok {
+			// port is already taken
+			continue
+		}
+
+		protocols, err := addRuleForPort(pt.PodPortRules, port, podIP, podPort, protocol)
+		if err != nil {
+			klog.ErrorS(err, "Port cannot be reserved, moving on to the next one", "port", port)
+			continue
+		}
+
+		pt.PortSearchStart = port + 1
+		if pt.PortSearchStart > pt.EndPort {
+			pt.PortSearchStart = pt.StartPort
+		}
+		return port, protocols, nil
+	}
+	return 0, nil, fmt.Errorf("no free port found")
+}
+
+func (pt *PortTable) AddRule(podIP string, podPort int, protocol string) (int, error) {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	npData := pt.getEntryByPodIPPortProto(podIP, podPort, protocol)
+	exists := (npData != nil)
+	if !exists {
+		nodePort, protocols, err := pt.addRuleforFreePort(podIP, podPort, protocol)
+		//success means port, protocol available, state is open.
+		if err != nil {
+			return 0, err
+		}
+		npData = &NodePortData{
+			NodePort:  nodePort,
+			PodIP:     podIP,
+			PodPort:   podPort,
+			Protocols: protocols,
+		}
+
+		pt.NodePortTable[nodePort] = npData
+		pt.PodEndpointTable[podIPPortProtoFormat(podIP, podPort, protocol)] = npData
+	} else {
+		protocolSocketData := npData.FindProtocol(protocol)
+		if protocolSocketData == nil {
+			return 0, fmt.Errorf("unknown protocol %s", protocol)
+		}
+		if protocolSocketData.State == stateInUse {
+			return 0, fmt.Errorf("rule for %s:%d:%s already exists", podIP, podPort, protocol)
+		}
+		if protocolSocketData.State == stateClosed {
+			return 0, fmt.Errorf("invalid closed state for %s:%d:%s", podIP, podPort, protocol)
+		}
+
+		if protocolSocketData.State == stateOpen {
+			return 0, fmt.Errorf("invalid open state for %s:%d:%s", podIP, podPort, protocol)
+		}
+
+	}
+	return npData.NodePort, nil
+}
+
+// syncRules ensures that contents of the port table matches the netnat rules present on the Node.
+func (pt *PortTable) syncRules() error {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	nplPorts := make([]rules.PodNodePort, 0, len(pt.NodePortTable))
+	for _, npData := range pt.NodePortTable {
+		protocols := make([]string, 0, len(supportedProtocols))
+		for _, protocol := range npData.Protocols {
+			if protocol.State == stateInUse {
+				protocols = append(protocols, protocol.Protocol)
+			}
+		}
+		nplPorts = append(nplPorts, rules.PodNodePort{
+			NodePort:  npData.NodePort,
+			PodPort:   npData.PodPort,
+			PodIP:     npData.PodIP,
+			Protocols: protocols,
```

Comment:
I think you should also modify this struct to support `Protocol` field, which is expected to be the same as the annotation?

Reply:
Updated PodNodePort struct.